### PR TITLE
docs: add simple example for network field

### PR DESCRIPTION
**Pull request by googs1025**
- Fix https://github.com/kubernetes-sigs/jobset/issues/548
I would like to add a usage example of the network field in simple/example and demonstrate how to access each other between two pods. This will allow users to better understand the functions provided by the jobset itself.

### CODE REVIEW

**Comment by googs1025** [comment]
I'm not sure if I need to add documentation demonstrating how to ping between two pods.
like this
```bash
root@VM-0-4-ubuntu:/home/ubuntu# kubectl get pods
NAME                               READY   STATUS      RESTARTS   AGE
failure-policy-workers-0-0-c59sq   0/1     Completed   0          20h
failure-policy-workers-0-1-98htc   0/1     Completed   0          20h
network-jobset-leader-0-0-5dmv7    1/1     Running     0          20m
network-jobset-workers-0-0-g5n2t   1/1     Running     0          20m
network-jobset-workers-0-1-rghv6   1/1     Running     0          20m
paralleljobs-driver-0-0-dw5mh      0/1     Completed   0          20h
paralleljobs-workers-0-0-mfwtk     0/1     Completed   0          20h
paralleljobs-workers-0-1-5g6gp     0/1     Completed   0          20h
paralleljobs-workers-0-2-7n9j5     0/1     Completed   0          20h
paralleljobs-workers-0-3-vp8jc     0/1     Completed   0          20h
success-policy-leader-0-0-8rwff    0/1     Completed   0          45h
success-policy-workers-0-0-c58gl   0/1     Completed   0          45h
success-policy-workers-0-1-dhfq9   0/1     Completed   0          45h
root@VM-0-4-ubuntu:/home/ubuntu# kubectl exec -it network-jobset-leader-0-0-5dmv7 -- sh
/ # cat /etc/hosts
# Kubernetes-managed hosts file.
127.0.0.1	localhost
::1	localhost ip6-localhost ip6-loopback
fe00::0	ip6-localnet
fe00::0	ip6-mcastprefix
fe00::1	ip6-allnodes
fe00::2	ip6-allrouters
10.6.2.23	network-jobset-leader-0-0.network-jobset.default.svc.cluster.local	network-jobset-leader-0-0
/ # ping network-jobset-workers-0-1.network-jobset.default.svc.cluster.local
PING network-jobset-workers-0-1.network-jobset.default.svc.cluster.local (10.6.2.24): 56 data bytes
64 bytes from 10.6.2.24: seq=0 ttl=63 time=0.069 ms
64 bytes from 10.6.2.24: seq=1 ttl=63 time=0.062 ms
64 bytes from 10.6.2.24: seq=2 ttl=63 time=0.066 ms
^C
--- network-jobset-workers-0-1.network-jobset.default.svc.cluster.local ping statistics ---
3 packets transmitted, 3 packets received, 0% packet loss
round-trip min/avg/max = 0.062/0.065/0.069 ms

```

File: README.md [hunk 1]
```diff
@@ -17,7 +17,7 @@ Take a look at the [concepts](https://jobset.sigs.k8s.io/docs/concepts/) page fo
...
+- **Automatic headless service configuration and lifecycle management**: ML and HPC frameworks require a stable network endpoint for each worker in the distributed workload, and since pod IPs are dynamically assigned and can change between restarts, stable pod hostnames are required for distributed training on k8s, By default, JobSet uses [IndexedJobs](https://kubernetes.io/blog/2021/04/19/introducing-indexed-jobs/) to establish stable pod hostnames, and does automatic configuration and lifecycle management of the headless service to trigger DNS record creations and establish network connectivity via pod hostnames.You can run this [example](examples/network-policy/jobset-with-network.yaml) yourself to see how exclusive placement works.
```
Comment (danielvegamyhre) [inline README.md:20]: This jobset-with-network.yaml example doesn't demonstrate exclusive placement. This is just showing how to use a custom subdomain.

Comment (googs1025) [reply README.md]: done

File: README.md [hunk 2]
```diff
@@ -17,7 +17,7 @@ Take a look at the [concepts](https://jobset.sigs.k8s.io/docs/concepts/) page fo
...
+- **Automatic headless service configuration and lifecycle management**: ML and HPC frameworks require a stable network endpoint for each worker in the distributed workload, and since pod IPs are dynamically assigned and can change between restarts, stable pod hostnames are required for distributed training on k8s, By default, JobSet uses [IndexedJobs](https://kubernetes.io/blog/2021/04/19/introducing-indexed-jobs/) to establish stable pod hostnames, and does automatic configuration and lifecycle management of the headless service to trigger DNS record creations and establish network connectivity via pod hostnames.You can run this [example](examples/network-policy/jobset-with-network.yaml) yourself to see how to use a custom subdomain.
```
Comment (danielvegamyhre) [inline README.md:20]: Let's rephrase to "These networking configurations are defaulted automatically to enable stable network endpoints and pod-to-pod communication via hostnames; however, they can be customized in the JobSet spec: see this example of using a custom subdomain your JobSet's network configuration."

**Review by danielvegamyhre — CHANGES_REQUESTED** [review]
left one comment

File: README.md (hunk 2 above)
Comment (googs1025) [reply README.md]: done

File: README.md [hunk 3]
```diff
@@ -17,7 +17,7 @@ Take a look at the [concepts](https://jobset.sigs.k8s.io/docs/concepts/) page fo
...
+- **Automatic headless service configuration and lifecycle management**: ML and HPC frameworks require a stable network endpoint for each worker in the distributed workload, and since pod IPs are dynamically assigned and can change between restarts, stable pod hostnames are required for distributed training on k8s, By default, JobSet uses [IndexedJobs](https://kubernetes.io/blog/2021/04/19/introducing-indexed-jobs/) to establish stable pod hostnames, and does automatic configuration and lifecycle management of the headless service to trigger DNS record creations and establish network connectivity via pod hostnames.These networking configurations are defaulted automatically to enable stable network endpoints and pod-to-pod communication via hostnames; however, they can be customized in the JobSet spec: see this [example](examples/network-policy/jobset-with-network.yaml) of using a custom subdomain your JobSet's network configuration.
```
Comment (danielvegamyhre) [inline README.md:20]: Missing a space after the period here: `...pod hostnames.These...`

File: examples/network-policy/README.md [hunk 4]
```diff
@@ -0,0 +1,48 @@
+## Usage example
+### Purpose
+This document provides an example of achieving communication between different pods using a headless service with JobSet.
+
+- apply and check whether the jobset's pod and svc start normally
```
Comment (danielvegamyhre) [inline examples/network-policy/README.md:5]: For documentation, please use full English sentences in documentation with proper capitalization, punctuation, spelling, etc.

**Review by danielvegamyhre — COMMENTED** [review]
Hmm I'm unsure if we want this network policy example in it's own folder, rather than in examples/simple. I guess you did this since you wanted to add a README explaining some extra steps the user can take to verify the pod-to-pod communication via pod hostnames is working correctly.

@kannon92 what do you think about this? The readme is a nice touch but I don't want the examples folder to get disorganized. Maybe we can think about each example having a short readme to accompany it like this?

File: examples/network-policy/README.md (hunk 4 above)
Comment (googs1025) [reply examples/network-policy/README.md:5]: I will make changes in the next few days. I am not a native English speaker, so I may need you to check after I complete the modification.

**Comment by googs1025** [comment]
> Hmm I'm unsure if we want this network policy example in it's own folder, rather than in examples/simple. I guess you did this since you wanted to add a README explaining some extra steps the user can take to verify the pod-to-pod communication via pod hostnames is working correctly.
> 
> @kannon92 what do you think about this? The readme is a nice touch but I don't want the examples folder to get disorganized. Maybe we can think about each example having a short readme to accompany it like this?

Yes, I separated them because I wanted to have a README.md document to describe .

**Comment by googs1025** [comment]
@danielvegamyhre @kannon92 After thinking for a while, I've decided to put it back into the examples/simple directory. How about this? I will move the README.md file to the Troubleshooting section [here](https://jobset.sigs.k8s.io/docs/troubleshooting/) because network connectivity issues are something that users should be aware of. If users are unsure, they can refer to the Troubleshooting section for guidance.

File: site/content/en/docs/troubleshooting/_index.md [hunk 5]
```diff
@@ -59,3 +59,49 @@ Look at the JobSet controller logs and you'll probably see an error like this:
 **Cause**: This could be due to a known bug in an older version of JobSet, or a known bug in an older version of Kueue. JobSet and Kueue integration requires JobSet v0.2.3+ and Kueue v0.4.1+.
 
 **Solution**: If you're using JobSet version less than v0.2.3, uninstall and re-install using a versoin >= v0.2.3 (see the JobSet [installation guide](https://jobset.sigs.k8s.io/docs/installation/) for the commands to do this). If you're using a Kueue version less than v0.4.1, uninstall and re-install using a v0.4.1 (see the Kueue [installation guide](https://kueue.sigs.k8s.io/docs/installation/) for the commands to do this).
+
+## 4. Using a headless service of JobSet to enable communication between different Pods
```
Comment (danielvegamyhre) [inline site/content/en/docs/troubleshooting/_index.md:63]:
```suggestion
## 4. Troubleshooting network communication between different Pods
```

File: site/content/en/docs/troubleshooting/_index.md [hunk 6]
```diff
@@ -59,3 +59,49 @@ Look at the JobSet controller logs and you'll probably see an error like this:
...
+
+**Solution**: First, we can deploy the example by running `kubectl apply -f jobset-network.yaml` [example](../../../../../site/static/examples/simple/jobset-with-network.yaml) and then check if the pods and services of the JobSet are running correctly. We can use the exec command to enter the container. By checking the /etc/hosts file within the container, we can observe the presence of a domain name, such as network-jobset-leader-0-0.example.default.svc.cluster.local. This domain name allows other containers to access the current pod. Similarly, we can utilize the domain names of other pods for network communication.
```
Comment (danielvegamyhre) [inline site/content/en/docs/troubleshooting/_index.md:65]: This is still phrased like a tutorial. Can we update the phrasing to indicate these are debugging steps?

Also please mention possible causes of networking issues

File: site/content/en/docs/troubleshooting/_index.md (hunk 5 above)
Comment (googs1025) [reply site/content/en/docs/troubleshooting/_index.md]: done

File: site/content/en/docs/troubleshooting/_index.md (hunk 6 above)
Comment (googs1025) [reply site/content/en/docs/troubleshooting/_index.md]: done

File: examples/simple/jobset-with-network.yaml [hunk 7]
```diff
@@ -0,0 +1,51 @@
...
+          template:
+            spec:
+              containers:
+                - name: leader
+                  image: bash:latest
+                  command:
+                    - bash
+                    - -xc
+                    - |
+                      sleep 3600
```
Comment (danielvegamyhre) [inline examples/simple/jobset-with-network.yaml:34]: You should have one of these pods ping another pod using <podHostname>.<subdomain> to show how the subdomain works.

File: site/static/examples/simple/jobset-with-network.yaml [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+          template:
+            spec:
+              containers:
+                - name: leader
+                  image: bash:latest
+                  command:
+                    - bash
+                    - -xc
+                    - |
+                      ping network-jobset-workers-0-.example.default.svc.cluster.local
```
[the remark this answers is not in the document]
Comment (googs1025) [reply site/static/examples/simple/jobset-with-network.yaml]: my bad. i change it

**Comment by googs1025** [comment]
@danielvegamyhre /PTAL Any other suggestions or modifications? thanks!

File: examples/simple/jobset-with-network.yaml [hunk 9]
```diff
@@ -0,0 +1,61 @@
...
+                - name: worker
+                  image: bash:latest
+                  command:
+                    - bash
+                    - -xc
+                    - |
+                      sleep 20
+                      success_count=0
+                      for i in {1..2}; do
+                        if ping -c 1 network-jobset-leader-0-0.example.default.svc.cluster.local; then
```
Comment (danielvegamyhre) [inline examples/simple/jobset-with-network.yaml:54]: This should just use the pod hostname `network-jobset-leader-0-0.example`

File: site/content/en/docs/troubleshooting/_index.md [hunk 10]
```diff
@@ -59,3 +59,50 @@ Look at the JobSet controller logs and you'll probably see an error like this:
...
+kubernetes   ClusterIP   10.96.0.1    <none>        443/TCP   2d1h
+```
+
+```bash
+root@VM-0-4-ubuntu:/home/ubuntu# kubectl exec -it network-jobset-leader-0-0-5xnzz -- sh
+/ # cat /etc/hosts
+# Kubernetes-managed hosts file.
+127.0.0.1	localhost
+...
+10.6.2.27	network-jobset-leader-0-0.example.default.svc.cluster.local	network-jobset-leader-0-0
```
Comment (danielvegamyhre) [inline site/content/en/docs/troubleshooting/_index.md:89]: Examples shouldn't use the full `network-jobset-leader-0-0.example.default.svc.cluster.local`, just use the pod hostname and service: `network-jobset-leader-0-0.example`

File: site/static/examples/simple/jobset-with-network.yaml [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+                - name: worker
+                  image: bash:latest
+                  command:
+                    - bash
+                    - -xc
+                    - |
+                      sleep 20
+                      success_count=0
+                      for i in {1..2}; do
+                        if ping -c 1 network-jobset-leader-0-0.example.default.svc.cluster.local; then
```
Comment (danielvegamyhre) [inline site/static/examples/simple/jobset-with-network.yaml:54]: Same here, just use pod hostname and service. Please check all the other places in this doc are updated

**Comment by danielvegamyhre** [comment]
/lgtm
/approve

Thanks @googs1025!

**Pull request merged**